### ISSUE TITLE
[🚀 Feature Request] pdf and sampling from Alpha-stable distribution 

### ISSUE THREAD

**Issue by jk4011** [pytorch/pytorch#91300]
### 🚀 The feature, motivation and pitch

I'm a long-time user of pytorch but a first-time contributor, and this is my first issue. I propose the implementation of the pdf and sampling of alpha-stable distribution into Pytorch. 

Recently, my team published a paper on the Neurips 2022 Workshop titled [[Score-Based Generative Models using Lévy Processes](https://openreview.net/forum?id=ErzyBArv6Ue)](https://openreview.net/forum?id=ErzyBArv6Ue). In this study, gaussian noise on diffusion model is replaced with  alpha stable noise. Alpha-stable distribution is unfortunately not supported by PyTorch. Therefore, I implemented this feature to a different repository named torchlevy. However, it will be preferable if the pdf and sample features are included in PyTorch so that other AI researchers may quickly and readily utilize them.

### How to implement

- Implementation of pdf and sampling method is based on Zolotarev thm([Stable Diffusion](https://papers.ssrn.com/sol3/Delivery.cfm/SSRN_ID2894444_code545.pdf?abstractid=2894444&mirid=1)], page 7.)
- I want to add new file on `torch/distribution` for these features.

### Issue

- Pdf caculation requires numerical integration which is not supported on Pytorch. In torchlevy, I’ve used another python module called torchquad for efficient numerical integration. But I’m not sure I can use this module on Pytorch. If I cannot use this module, I want to implement only sampling from alpha stable distribution.

cc [@fritzo](https://github.com/fritzo) [@neerajprad](https://github.com/neerajprad) [@alicanb](https://github.com/alicanb) [@nikitaved](https://github.com/nikitaved)

### Additional context

I would be happy to work on solving this when the issue has been discussed further.

**Comment by fritzo** [comment]
Is your torchlevy repo public?  I currently get a 404 not found error on github.

Note Pyro implements a [Stable distribution](https://docs.pyro.ai/en/stable/distributions.html#pyro.distributions.Stable) with reparametrized sampler but no density.  In most cases where a density is needed, we instead use [auxiliary variable methods](https://docs.pyro.ai/en/stable/infer.reparam.html#pyro.infer.reparam.stable.StableReparam) to convert a single Stable random variable to a few Gaussian random variables plus some nonlinear transformations.  Here's [a tutorial](https://pyro.ai/examples/forecasting_i.html) demonstrating the interface.

If we do add a `Stable` distribution to torch.distributions, I'd recommend using Nolan's S0 parametrization because that coordinate system supports reparametrized sampling aka pathwise derivatives.  By constrast the S parameterization used by scipy is discontinuous at α=1 and has poor geometry for pathwise derivatives.

> Pdf caculation requires numerical integration

I wonder if you could instead provide an unbiased Monte Carlo estimate of `.log_prob()`, which would be enough to train diffusion models 🤔

**Comment by fritzo** [comment]
Duplicate of https://github.com/pytorch/pytorch/issues/19248

**Comment by jk4011** [comment]
> Is your torchlevy repo public? I currently get a 404 not found error on github.
I'm sorry. torchlevy is not public yet since anonymous policy of conference on the review. But I'll open it in few weeks.

> I wonder if you could instead provide an unbiased Monte Carlo estimate of .log_prob(), which would be enough to train diffusion models
Pdf calculation is not used in training. It is used calculating likelihood of intractable data distribution in which diffusion model want to imitate. 

I used Simpson's rules for numerical integration with range [-10, 10]. Unfortunately, I didn't care so much about biasness of estimation. If you don't mind, could you explain what kind of biasness I can encounter when I use numerical integration for pdf calculation?

**Comment by jk4011** [comment]
> Is your torchlevy repo public? I currently get a 404 not found error on github.

I'm sorry. torchlevy is not public yet since anonymous policy of conference on the review. But I'll open it in couple of months.

> I wonder if you could instead provide an unbiased Monte Carlo estimate of .log_prob(), which would be enough to train diffusion models

Pdf calculation is not used in training. It is used calculating likelihood of intractable data distribution in which diffusion model imitates. 

There is also the option to use direct numerical integration of the standard parameterization of the characteristic function or to evaluate by taking the FFT of the characteristic function.

cc @fritzo

**Comment by fritzo** [comment]
>  While I am unsure [about] biases in the PDF calculation

Oh I'm assuming you're implementation is correct and it would be great to have it in torch.distributions once we can remove external dependencies 😄

I'm just wondering about speed-accuracy tradeoffs.  In the experience @martinjankowiak and I had with heavy-tailed distributions in PyTorch, our Stable implementation led to end-to-end performance about as good as a StudentT distribution, but was about 4x slower, so we couldn't recommend Stable over StudentT.  And I'm guessing there's significant speed penalty to a `.log_prob()` method based on quadrature or FFT.  I bet an fft-based `Stable.log_prob()` is >100x more expensive than `StudentT.log_prob()`.  So I was just thinking, hey since most applications of PyTorch use stochastic gradient learning, we don't really need an exact `.log_prob()` computation, rather would be enough if we could compute a cheap stochastic estimate of `.log_prob()`, one that agrees with the `.log_prob()` in expectation (i.e. is unbiased) but is faster than computing an exact log density.  This is just an off-the cuff idea, I don't see an easy way to express `.log_prob()` as an integral amenable to Monte Carlo sampling, instead we have an integral representation of the density (`.log_prob().exp()` rather than the log-density `.log_prob()`).

In summary:
- `Stable.log_prob()` has historically been expensive to compute
- PyTorch applications typically require less than a log-density computation, merely a stochastic unbiased estimate of log-density
- Maybe there's an opportunity to relax interface requirements and build a cheaper but useful distribution?

**Comment by jk4011** [comment]
Thank you for your comment!
> And I'm guessing there's significant speed penalty to a `.log_prob()` method based on quadrature or FFT. I bet an fft-based `Stable.log_prob()` is >100x more expensive than `StudentT.log_prob()`.
> 

Yes, fft-based `Stable.log_prob()` is >100x more expensive than exact form pdf. I’m very sad stable distribution is not expressed exact form. Despite this limitation, the stable distribution is still widely used in various tasks, including my research. While Scipy does support for calculating the pdf, it is super heavy. On the other hand, PyTorch can be better option for pdf calculation because of its comfortable parallelization interface.

**Comment by jk4011** [comment]
> So I was just thinking, hey since most applications of PyTorch use stochastic gradient learning, we don't really need an exact `.log_prob()` computation, rather would be enough if we could compute a cheap stochastic estimate of `.log_prob()`, one that agrees with the `.log_prob()` in expectation (i.e. is unbiased) but is faster than computing an exact log density.
> 

I have been using linear interpolation to approximate the probability density function (pdf) of the stable distribution using cached pdf values spaced 0.01 units apart over the range of -100 to 100. Because the pdf is a one-dimensional function, this method has proven effective in providing a fairly accurate approximation using a small number of cached values. However, I am unsure if this approach can be implemented within the PyTorch framework. Do you think this would be possible?

**Comment by jk4011** [comment]
> This is just an off-the cuff idea, I don't see an easy way to express `.log_prob()` as an integral amenable to Monte Carlo sampling, instead we have an integral representation of the density (`.log_prob().exp()` rather than the log-density `.log_prob()`).
> 

Currently, we can think of first calculating pdf and taking the logarithm. However, this approach can lead to inaccuracies when the |x| is large. I've been thinking about this problem, but I haven't found a good alterative to solve it.
